pkg/sql/catalog/descbuilder/desc_builder.go

```diff
@@ -21,39 +22,151 @@ import (
+) (catalog.DescriptorBuilder, error) {
+	b := newBuilder(descProto, mvccTimestamp)
+	if b == nil {
+		return nil, nil
+	}
+	if err := b.RunPostDeserializationChanges(); err != nil {
+		return nil, err
+	}
+	return b, nil
+}
+
+func newBuilder(
+	descProto *descpb.Descriptor, mvccTimestamp hlc.Timestamp,
 ) catalog.DescriptorBuilder {
-	table, database, typ, schema, function := descpb.FromDescriptorWithMVCCTimestamp(desc, mvccTimestamp)
-	switch {
-	case table != nil:
-		return tabledesc.NewBuilder(table)
-	case database != nil:
-		return dbdesc.NewBuilder(database)
-	case typ != nil:
-		return typedesc.NewBuilder(typ)
-	case schema != nil:
-		return schemadesc.NewBuilder(schema)
-	case function != nil:
-		return funcdesc.NewBuilder(function)
-	default:
-		return nil
+	switch t := descProto.Union.(type) {
+	case *descpb.Descriptor_Table:
+		return tabledesc.NewBuilderWithMVCCTimestamp(t.Table, mvccTimestamp)
+	case *descpb.Descriptor_Database:
+		return dbdesc.NewBuilderWithMVCCTimestamp(t.Database, mvccTimestamp)
+	case *descpb.Descriptor_Type:
+		return typedesc.NewBuilderWithMVCCTimestamp(t.Type, mvccTimestamp)
+	case *descpb.Descriptor_Schema:
+		return schemadesc.NewBuilderWithMVCCTimestamp(t.Schema, mvccTimestamp)
+	case *descpb.Descriptor_Function:
+		return funcdesc.NewBuilderWithMVCCTimestamp(t.Function, mvccTimestamp)
 	}
+	return nil
 }
 
-// NewBuilder is a convenience function which calls NewBuilderWithMVCCTimestamp
-// with an empty timestamp.
-// The expectation here, therefore, is that this function is called either for a
-// new descriptor that doesn't exist in storage yet, or already has its
-// modification time field (and others which depend on the MVCC timestamp)
-// set to a valid value.
-func NewBuilder(desc *descpb.Descriptor) catalog.DescriptorBuilder {
-	return NewBuilderWithMVCCTimestamp(desc, hlc.Timestamp{})
+// BuildMutable generates a catalog.MutableDescriptor from scratch based on:
+//   - the original descriptor (optional),
+//   - the mutated descriptor,
+//   - an MVCC timestamp (optional) to set the mutated descriptor's
+//     ModificationTime with.
+//
+// The ID and Version fields may also be overwritten for correctness in case
+// they deviate from what's expected.
+func BuildMutable(
+	original catalog.Descriptor, mutated *descpb.Descriptor, mvccTimestamp hlc.Timestamp,
+) (catalog.MutableDescriptor, error) {
+	if mutated == nil {
+		return nil, errors.AssertionFailedf("empty mutated descriptor")
+	}
+	pb := protoutil.Clone(mutated).(*descpb.Descriptor)
+	correctFieldValues := func(newID *descpb.ID, newVersion *descpb.DescriptorVersion, newModTime *hlc.Timestamp) {
+		if original != nil && *newID != original.GetID() {
+			*newID = original.GetID()
+		}
+		if original == nil {
+			*newVersion = 1
+		} else if *newVersion != original.GetVersion() {
+			*newVersion = original.GetVersion()
+		}
+		if !mvccTimestamp.IsEmpty() && newModTime.IsEmpty() {
+			*newModTime = mvccTimestamp
+		}
+	}
```

Comment:
I trust you on this. But I'd like to have some explanation why this kind correction is ok. Specifically, how come descriptor ID can be different and why that is "fine"?